### PR TITLE
feat: add occurence information to defs and inline defs that occur once

### PR DESCRIPTION
**Pull request by ChrBondi**

### CODE REVIEW

**Comment by magnus-madsen** [comment]
> @magnus-madsen We seems to have a problem with converting tail calls. We are getting errors with the stack frame and we are not sure if `convertSelfTailCall` and `convertTailCall` are correct

If it mirrors `Tailrec` then it should be OK. Right? You can always try to disable that phase for experimentation.

**Comment by ChrBondi** [comment]
It does mirror `TailRec` except we had to add an extra case for ApplySelfTail. We still get the problem if we disable TailRec

**Comment by magnus-madsen** [comment]
> It does mirror `TailRec` except we had to add an extra case for ApplySelfTail. We still get the problem if we disable `TailRec`

If you disable Tailrec there should be no Apply*Tail right? So then the problem must lie elsewhere?

**Comment by ChrBondi** [comment]
It seems like the problem occurs if the inlined function is too large. It works if we set the threshold lower but we can't really argue why this is correct

**Comment by magnus-madsen** [comment]
> It seems like the problem occurs if the inlined function is too large. It works if we set the threshold lower but we can't really argue why this is correct

How large are we talking? The JVM has various limits.

**Comment by ChrBondi** [comment]
We are counting the amount of expressions in the body of a def and setting a treshold that where we only inline defs that are below this. A threshold of 8 works. A threshold of 16 breaks

**Comment by magnus-madsen** [comment]
> We are counting the amount of expressions in the body of a def and setting a treshold that where we only inline defs that are below this. A threshold of 8 works. A threshold of 16 breaks

Not sure I can help much. That just sounds strange.

**Comment by magnus-madsen** [comment]
Let me know when I should review.

**Comment by ChrBondi** [comment]
@magnus-madsen should be ready now

**Comment by magnus-madsen** [comment]
I did a quick review. Once that is done, I will do a deeper review.

**Comment by magnus-madsen** [comment]
What's the performance like?

File: main/src/ca/uwaterloo/flix/language/phase/Inliner.scala [hunk 1]
```diff
@@ -497,10 +514,66 @@ object Inliner {
   /**
    * Convert a given tailCall expression `exp0` to a non tail call
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Inliner.scala:515]: => Converts all tail calls in `exp0` to non-tail calls.

File: main/src/ca/uwaterloo/flix/language/phase/Inliner.scala [hunk 2]
```diff
@@ -497,10 +514,66 @@ object Inliner {
...
    */
-  //TODO expand `convertTailCall` when more functions are being inlined
   private def convertTailCall(exp0: OccurrenceAst.Expression): OccurrenceAst.Expression = exp0 match {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Inliner.scala:517]: convertTailCall -> rewriteTailCalls (or maybe removeTailCalls)

File: main/src/ca/uwaterloo/flix/language/phase/Inliner.scala [hunk 3]
```diff
@@ -444,6 +451,16 @@ object Inliner {
     case OccurrenceAst.Expression.MatchError(tpe, loc) => LiftedAst.Expression.MatchError(tpe, loc)
   }
 
+  /**
+   * A def can be inlined if
+   * Its body consist of a single non self call or
+   * Its body has a codesize below the inline threshold and its not being inlined into itself or
+   * It only occurs once in the entire program
+   */
+  private def canInlineDef(def0: OccurrenceAst.Def, sym0: Symbol.DefnSym): Boolean = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Inliner.scala:460]: sym0 argument should be first, by convention.

File: main/src/ca/uwaterloo/flix/language/phase/Inliner.scala [hunk 4]
```diff
@@ -497,10 +514,66 @@ object Inliner {
...
+
+    case OccurrenceAst.Expression.ApplySelfTail(sym, _, actuals, tpe, purity, loc) => OccurrenceAst.Expression.ApplyDef(sym, actuals, tpe, purity, loc)
+
+    case _ => exp0
+  }
+
+  /**
+   * Convert a given ApplySelfTail expression `exp0` to a ApplyDefTail
+   */
+  private def convertSelfTailCall(exp0: OccurrenceAst.Expression): OccurrenceAst.Expression = exp0 match {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Inliner.scala:552]: I am a bit spooked by this. 

Do you understand what this does? It seems to me this corresponds to unrolling a recursive loop once. Is that what you want? Is that "good"?

File: main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala [hunk 5]
```diff
@@ -19,33 +19,44 @@ package ca.uwaterloo.flix.language.phase
...
 /**
- * The occurrence analyzer collects information on variable usage
- * Marks a variable as Dead if it is not used, Once if it is used exactly once and Many otherwise
+ * The occurrence analyzer collects information on variable and function usage and calculates the weight of the expressions
+ * Marks a variable or function as Dead if it is not used, Once if it is used exactly once and Many otherwise
  */
 object OccurrenceAnalyzer {
 
+  object OccurInfo {
+    val EmptyOccur: OccurInfo = OccurInfo(Map.empty, Map.empty, 0)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala:36]: Add comments (i.e. when would you use this?)

File: main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala [hunk 6]
```diff
@@ -19,33 +19,44 @@ package ca.uwaterloo.flix.language.phase
...
+
+    val BaseOccur: OccurInfo = OccurInfo(Map.empty, Map.empty, 1)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala:38]: Add comments (i.e. when would you use this?)

File: main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala (hunk 5 above)
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala:36]: EmptyOccur -> Empty

File: main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala (hunk 6 above)
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala:38]: BaseOccur -> "One"

File: main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala [hunk 7]
```diff
@@ -19,33 +19,44 @@ package ca.uwaterloo.flix.language.phase
...
+  }
+
+  case class OccurInfo(varOccurrence: Map[VarSym, OccurrenceAst.Occur], defOccurrence: Map[DefnSym, OccurrenceAst.Occur], codeSize: Int) {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala:41]: Add documentation.

Usually these classes are sorted, i.e. defOccurrence would be before varOccurrence. You could also consider calling these just `defs` and `vars`.

File: main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala [hunk 8]
```diff
@@ -103,89 +129,107 @@ object OccurrenceAnalyzer {
   /**
    * Performs occurrence analysis on the given expression `exp0`
    */
-  private def visitExp(exp0: LiftedAst.Expression): (OccurrenceAst.Expression, Map[Symbol.VarSym, OccurrenceAst.Occur]) = exp0 match {
-    case Expression.Unit(loc) => (OccurrenceAst.Expression.Unit(loc), Map.empty)
+  private def visitExp(exp0: LiftedAst.Expression)(implicit sym0: Symbol.DefnSym): (OccurrenceAst.Expression, OccurInfo) = exp0 match {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala:132]: symbol should not be implicit, but be the first argument.

File: main/src/ca/uwaterloo/flix/language/ast/OccurrenceAst.scala [hunk 9]
```diff
@@ -295,8 +295,10 @@ object OccurrenceAst {
    * `OccurDef` contains information that indicates whether or not a def should be inlined
    *  A def is `isTrivialNonSelfCall` if
    *  the expression consist of a single (non-self) call with trivial arguments
+   *  `occur` represents the amount of time a def is called in the entire program
+   *  `codeSize` denotes the cumulative weight of each expression in the body of the def
    */
-  case class DefContext(isNonSelfCall: Boolean)
+  case class DefContext(isNonSelfCall: Boolean, occur: Occur, codeSize: Int)
```
[the remark this answers is not in the document]
Comment (ChrBondi) [reply main/src/ca/uwaterloo/flix/language/ast/OccurrenceAst.scala:301]: It represents whatever the body of a def consist only of a reference to another def. In which case we can inline it by simply replacing one function call with another

[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/OccurrenceAst.scala:301]: Ok. How about we call it `isDirectCall` or `isDirectForwardingCall` or something like this?

File: main/src/ca/uwaterloo/flix/language/phase/Inliner.scala (hunk 4 above)
Comment (ChrBondi) [reply main/src/ca/uwaterloo/flix/language/phase/Inliner.scala:552]: It was used when inlining a self recursive function into the body of another function. 
```
def foo(x: Int32): Int32 = 
    if (x > 5)
        1
    else
        foo(x-1)

def bar(x: Int32): Int32 = foo(x)
```

If we wanted to inline foo into bar we would have to convert the `SelfTailCall` in foo into a `ApplyDefTail`

But this is not something we inline right now

**Comment by ChrBondi** [comment]
> What's the performance like?

The entire optimizer takes around two seconds more than before

**Comment by magnus-madsen** [comment]
> > What's the performance like?
> 
> The entire optimizer takes around two seconds more than before

Why? That seems like a lot!

**Comment by magnus-madsen** [comment]
Can you paste the output of benchmark phases?

**Comment by magnus-madsen** [comment]
and benchmark throughput with/without the optimizer?

**Comment by ChrBondi** [comment]
> Can you paste the output of benchmark phases?

<html>
<body>
<google-sheets-html-origin><meta name="sheets-banding" data-sheets-range="N7:P34" data-sheets-banding-properties="{&quot;1&quot;:{&quot;1&quot;:{&quot;4&quot;:{&quot;1&quot;:2,&quot;2&quot;:13411959}},&quot;2&quot;:{&quot;4&quot;:{&quot;1&quot;:2,&quot;2&quot;:16777215}},&quot;3&quot;:1,&quot;4&quot;:{&quot;4&quot;:{&quot;1&quot;:2,&quot;2&quot;:16315115}},&quot;5&quot;:1}}">

Finished 10 iterations on 54.590 lines of code in 161 seconds. |   |  
-- | -- | --
  |   |  
Best runtime per phase: |   |  
Phase | In ms | Percentage
EntryPoint | 0 | 0%
Deriver | 2,1 | 0%
Reader | 1,7 | 0%
Safety | 13,4 | 0,10%
Instances | 18,3 | 0,20%
Tailrec | 12,7 | 0,10%
Stratifier | 23,6 | 0,20%
VarNumbering | 18,4 | 0,20%
PatternExhaustiveness | 29,1 | 0,30%
Eraser | 23,7 | 0,20%
Simplifier | 35,2 | 0,30%
Optimizer | 751,3 | 7%
LambdaLift | 45,2 | 0,40%
Lowering | 51,5 | 0,50%
Kinder | 129,9 | 1,20%
ClosureConv | 146,2 | 1,40%
TreeShaker | 90 | 0,80%
Namer | 107,4 | 1%
Redundancy | 204,9 | 1,90%
Finalize | 135,1 | 1,30%
Weeder | 181,5 | 1,70%
Resolver | 288,2 | 2,70%
Parser | 589,1 | 5,50%
Monomorph | 1348,8 | 12,60%
JvmBackend | 2215,6 | 20,70%
Typer | 4185,6 | 39,20%
Regions | 39,1 | 0,40%


</body>
</html>

**Comment by ChrBondi** [comment]
> and benchmark throughput with/without the optimizer?

With the optimizer

Throughput (best):  4.341 lines/sec (with 8 threads.)

  min:  1.522, max:  4.341, avg:  3.285, median:  3.541

  The highest throughput was in iteration: 5 (out of 10).
  The ratio between the best and worst iteration was: 2,9x.

Finished 10 iterations on 54.606 lines of code in 181 seconds.


Without the optimizer

Throughput (best):  5.903 lines/sec (with 8 threads.)

  min:  1.488, max:  5.903, avg:  4.400, median:  4.959

  The highest throughput was in iteration: 3 (out of 10).
  The ratio between the best and worst iteration was: 4,0x.

Finished 10 iterations on 54.606 lines of code in 143 seconds.

**Comment by magnus-madsen** [comment]
This is a very big slow down. I think that still indicates something is "wrong". How much is spent in each subphase of the Optimizer?

**Comment by magnus-madsen** [comment]
If you set inline threshold to 10000000 does performance improve or not?

**Comment by magnus-madsen** [comment]
Observation: Are your def and var maps total? E.g. can they map variables/defs to dead? Or do you implicitly say store "Dead" by not having them in the map?

**Comment by magnus-madsen** [comment]
I seem to have been confused. This PR does not actually make the optimizer slower than it already was...

**Pull request merged**